common/scala/src/main/scala/whisk/core/database/MultipleReadersSingleWriterCache.scala

```diff
@@ -0,0 +1,313 @@
+import scala.util.Failure
+import scala.util.Success
+
+import spray.caching.Cache
+import spray.caching.LruCache
+import spray.caching.ValueMagnet.fromAny
+import whisk.common.Logging
+import whisk.common.LoggingMarkers
+import whisk.common.TransactionId
+
+/**
+ * A cache that allows multiple readers, bit only a single writer, at
+ * a time. It will make a best effort attempt to coalesce reads, but
+ * does not guarantee that all overlapping reads will be coalesced.
+ *
+ * The cache operates by bracketing all reads and writes. A read
+ * imposes a lightweight read lock, by inserting an entry into the
+ * cache with State.ReadInProgress. A write does the same, with
+ * State.WriteInProgress.
+ *
+ * On read or write completion, the value transitions to
+ * State.Cached.
+ *
+ * State.Initial is represented implicitly by absence from the cache.
+ *
+ * The handshake for cache entry state transition is:
+ * 1. if entry is in an agreeable state, proceed
+ *    where agreeable for reads is Initial, ReadInProgress, or Cached
+ *                    and the read proceeds by ensuring the entry is State.ReadInProgress
+ *      and agreeable for writes is Initial or Cached
+ *                    and the write proceeds by ensuring the entry is State.WriteInProgress
+ *
+ * 2. if entry is not in an agreeable state, fail fast with Future.failed(new ConcurrentUpdateException)
+ *
+ * 3. to swap in the new state to an existing entry, we use an AtomicReference.compareAndSet
+ *
+ * 4. after adding a new entry, we double check that the state of the actually-added entry
+ *    is agreeable with the desired action (as per item 1)
+ *
+ */
+trait MultipleReadersSingleWriterCache[W, Winfo] {
+    /** Subclasses: Toggle this to enable/disable caching for your entity type. */
+    protected def cacheEnabled = true
+
+    /** Subclasses: tell me what key to use for updates */
+    protected def cacheKeyForUpdate(w: W): Any
+
+    /**
+     * Each entry has a state
+     *
+     */
+    protected object State extends Enumeration {
+        type State = Value
+        val ReadInProgress, WriteInProgress, InvalidateInProgress, Cached = Value
+    }
+    import State._;
+    implicit def state2Atomic(state: State): AtomicReference[State] = {
+        new AtomicReference(state);
+    }
+
+    /**
+     * Failure modes
+     *
+     */
+    private class ConcurrentUpdateException extends Exception {}
+
+    /**
+     * The entries in the cache will be a pair of State and the Future value
+     *
+     */
+    private case class Entry(transid: TransactionId, state: AtomicReference[State], value: Future[W]) {
+        def invalidate {
+            state.set(InvalidateInProgress)
+        }
+
+        def writeDone()(implicit logger: Logging): Boolean = {
+            logger.debug(this, "Write finished")
+            trySet(WriteInProgress, Cached)
+        }
+
+        def readDone(value: W, promise: Promise[W])(implicit logger: Logging): Unit = {
```

Comment:
Like `writeDone` this should return a `Boolean` to be able to handle state further above and not needing to pass `Promise` around (mutability)